### DISCUSSION TITLE
Creating a Docker Image

### DISCUSSION THREAD

**Discussion by ricardo-noyolalozano** [testcontainers/testcontainers-dotnet#897] [Q&A] [answered]
Hey y'all, I'm having some trouble creating a docker image for my test. Not doing anything crazy yet, just trying to run my tests on my computer. This is the command I am using to create the image:

```
var image = new DockerImage("", "fakeservice", "test");
       
 _ = new ImageFromDockerfileBuilder()
            .WithDockerfileDirectory(CommonDirectoryPath.GetSolutionDirectory(), @"src\FakeService")
            .WithDockerfile("Dockerfile")
            .WithName(image)
            .Build()
            .CreateAsync()
            .ConfigureAwait(false);
```

This is the folder structure I have:
```
fakeapp/
├── src/
│   └── FakeService/
│      ├── FakeService.csproj
│      ├── Dockerfile
│      └── .dockerignore
├── tests/
│   └── Tests/
│       ├── Tests.csproj
│       └── MyTests.cs
└── FakeApp.sln
```

Calling the image:
```
_appContainer = new ContainerBuilder()
            .WithImage(image)
            .WithNetwork(_network)
            .Build();
```

Is there anything blatant I am doing wrong or what could be the issue I am having? When I specify a image I already have in the .WithImage call it works so I'm doing something wrong with creating the image.

**Comment by HofmeisterAn** [comment]
I do not see an obvious misconfiguration. It seems like the image building is failing. We might need to make the error messages more visible (Docker.DotNet does not throw an exception). Could you please share your reproducer project? I am happy to look into it. Otherwise, you can set a custom logger (`TestcontainersSettings.Logger`) that forwards all error messages or debug into:

https://github.com/testcontainers/testcontainers-dotnet/blob/e27f537ab6bf8b7c1a0321ed992cafa18a814ddb/src/Testcontainers/Clients/TraceProgress.cs#L20-L38

**Reply by ricardo-noyolalozano** [reply]
I'm not very tech savvy so not entirely sure how to set up the logger. Was able to get the image created by testing out a hunch of mine. I was misunderstanding what the documentation said about the 'tarball' and finally chased it down to it meaning that it creates that tar file with all the bin artifacts. My dockerfile was in context of it being executed in the main folder so it was lacking a lot of the resources it needed. I got the image to create correctly but now running into a 'System.Threading.Tasks' exception that I'll figure out.

**Reply by HofmeisterAn** [reply]
> My dockerfile was in context of it being executed in the main folder so it was lacking a lot of the resources it needed.

Great, you figured it out. That's what usually happens.

> I got the image to create correctly but now running into a 'System.Threading.Tasks' exception that I'll figure out.

If you can share the stack trace, I can take a look at it. Maybe I can point in the right direction.

**Reply by ricardo-noyolalozano** [reply]
Tests.FakeTests.SuccessfulConnection
   Source: FakeTests.cs line 83
   Duration: 1 ms

  Message: 
System.Threading.Tasks.TaskCanceledException : A task was canceled.

  Stack Trace: 
StreamUtil.MonitorStreamForMessagesAsync[T](Task`1 streamTask, DockerClient client, CancellationToken cancellationToken, IProgress`1 progress)
StreamUtil.MonitorResponseForMessagesAsync[T](Task`1 responseTask, DockerClient client, CancellationToken cancel, IProgress`1 progress)
DockerImageOperations.BuildAsync(IImageFromDockerfileConfiguration configuration, CancellationToken ct)
FutureDockerImage.UnsafeCreateAsync(CancellationToken ct)
FutureDockerImage.CreateAsync(CancellationToken ct)
FakeTests.InitializeAsync() line 59

It created the image once but now it seems that something is timing out and I can't get it to create the image again.

**Reply by ricardo-noyolalozano** [reply]
seems like it was the cancellation token that I was handing the .CreateAsync on the IFutureDockerImage that was timing out after a minute, but this is the stack trace after I remove it:

HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
DockerClient.PrivateMakeRequestAsync(TimeSpan timeout, HttpCompletionOption completionOption, HttpMethod method, String path, IQueryString queryString, IDictionary`2 headers, IRequestContent data, CancellationToken cancellationToken)
DockerClient.MakeRequestAsync(IEnumerable`1 errorHandlers, HttpMethod method, String path, IQueryString queryString, IRequestContent body, IDictionary`2 headers, TimeSpan timeout, CancellationToken token)
INetworkOperations.CreateNetworkAsync(NetworksCreateParameters parameters, CancellationToken cancellationToken)
DockerNetworkOperations.CreateAsync(INetworkConfiguration configuration, CancellationToken ct)
DockerNetwork.UnsafeCreateAsync(CancellationToken ct)
DockerNetwork.CreateAsync(CancellationToken ct)
FakeTests.InitializeAsync() line 49

**Reply by ricardo-noyolalozano** [reply, marked as answer]
It was all a timeout issue. Just needed to adjust the time on my CancellationToken. I am curious though the tests seem to be taking a long time. Is it creating the image every single time a test runs if I have the image creation in InitializeAsync()?

**Reply by HofmeisterAn** [reply]
> Is it creating the image every single time a test runs if I have the image creation in InitializeAsync()?

That's how xUnit.net's [shared context](https://xunit.net/docs/shared-context) work. Either implement the image build as a singleton or use a class or collection fixture (see the linked xUnit.net's docs). Maybe the [WeatherForecast](https://github.com/testcontainers/testcontainers-dotnet/blob/f0fef3ee3346665842426f1fe4bdead322c0097e/examples/WeatherForecast/tests/WeatherForecast.Tests/WeatherForecastImage.cs) example helps too, it uses a similar configuration and builds the image once.

**Reply by ricardo-noyolalozano** [reply]
Thank you
